src/Compatibility/Microsoft.DotNet.ApiCompatibility/Rules/AssemblyIdentityMustMatch.cs

```diff
@@ -0,0 +1,98 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Text;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+
+namespace Microsoft.DotNet.ApiCompatibility.Rules
+{
+    public class AssemblyIdentityMustMatch : Rule
+    {
+        public override void Initialize(RuleRunnerContext context)
+        {
+            context.RegisterOnAssemblySymbolAction(RunOnAssemblySymbol);
+        }
+
+        private void RunOnAssemblySymbol(IAssemblySymbol left, IAssemblySymbol right, string leftName, string rightName, IList<CompatDifference> differences)
+        {
+            if (left == null && right != null)
+            {
+                differences.Add(new CompatDifference(DiagnosticIds.AssemblyIdentityMustMatch, string.Format(Resources.AssemblyNameDoesNotExist, leftName, right.Identity.Name), DifferenceType.Removed, right.Identity.GetDisplayName()));
+                return;
+            }
+
+            if (left != null && right == null)
+            {
+                differences.Add(new CompatDifference(DiagnosticIds.AssemblyIdentityMustMatch, string.Format(Resources.AssemblyNameDoesNotExist, rightName, left.Identity.Name), DifferenceType.Added, left.Identity.GetDisplayName()));
+                return;
+            }
+
+            string leftAssemblyName = left.Identity.Name;
```

Comment:
Is it worth checking for the name? If the assembly name doesn't match, the mapper won't map them as the same assembly, what I'm trying to say is that when this rule runs, if both left and right != null, that means they have the same name.

Reply:
Never mind, with the entrypoints for single assemblies, this could happen.